File: core/src/com/inkostilation/pong/processing/Serializer.java
package com.inkostilation.pong.processing;

import com.google.gson.*;
import com.google.gson.reflect.TypeToken;
import com.inkostilation.pong.commands.ICommand;

import java.lang.reflect.Type;

public class Serializer {

    private static Gson gson;

    public Serializer() {
        GsonBuilder builder = new GsonBuilder();
        builder.registerTypeAdapter(Parcel.class, new ParcelDeserializer());
        gson = builder.create();
    }

    public ICommand deserialize(String json) {
        try {
            Parcel parcel = gson.fromJson(json, Parcel.class);
            Class<? extends ICommand> commandClass = null;
            commandClass = (Class<? extends ICommand>) Class.forName(parcel.commandClass);
            return commandClass.cast(parcel.command);
        } catch (ClassNotFoundException | JsonParseException e) {
            e.printStackTrace();
            return null;
        }
    }

    public String serialize(ICommand command) {
        return gson.toJson(new Parcel(command), Parcel.class);
    }

    private final class Parcel {
        String commandClass;
        ICommand command;

        public Parcel(ICommand command) {
            this.command = command;
            commandClass = command.getClass().getCanonicalName();
        }
    }

    private final class ParcelDeserializer implements JsonDeserializer<Parcel> {

        @Override
        public Parcel deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            final JsonObject jsonObject = json.getAsJsonObject();
            final JsonPrimitive commandClass = (JsonPrimitive) jsonObject.get("commandClass");
            if (commandClass == null) {
                throw new JsonParseException("Incorrect Parcel object: \"commandClass\" field not found. ");
            }
            final String className = commandClass.getAsString();
            final JsonElement commandJson = jsonObject.get("command");
            ICommand command;
            try {
                command = context.deserialize(commandJson, Class.forName(className));
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
                command = null;
            }
            return new Parcel(command);
        }
    }
}

File: core/src/com/inkostilation/pong/processing/Serializer.kt
package com.inkostilation.pong.processing

import com.google.gson.*
import com.inkostilation.pong.commands.ICommand
import java.lang.reflect.Type
import kotlin.jvm.Throws

class Serializer {
    fun deserialize(json: String): ICommand<*, *> {
        val parcel = gson.fromJson(json, Parcel::class.java)
        val commandClass = Class.forName(parcel.commandClass) as Class<out ICommand<*, *>>
        return commandClass.cast(parcel.command)!!
    }

    fun deserialize(jsons: List<String>): List<ICommand<*, *>> {
        return jsons.map { deserialize(it) }
    }

    fun serialize(command: ICommand<*, *>): String {
        return gson.toJson(Parcel(command), Parcel::class.java)
    }

    fun serialize(commands: List<ICommand<*, *>>): List<String> {
        return commands.map { serialize(it) }
    }

    private class Parcel(var command: ICommand<*, *>) {
        var commandClass: String = command.javaClass.canonicalName
    }

    private class ParcelDeserializer : JsonDeserializer<Parcel> {
        @Throws(JsonParseException::class)
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Parcel {
            val jsonObject = json.asJsonObject
            val commandClass = jsonObject["commandClass"] as JsonPrimitive
            val className = commandClass.asString
            val commandJson = jsonObject["command"]
            val command: ICommand<*, *>
            command = context.deserialize<ICommand<*, *>>(commandJson, Class.forName(className))
            return Parcel(command)
        }
    }

    companion object {
        private val gson by lazy {
            val builder = GsonBuilder()
            builder.registerTypeAdapter(Parcel::class.java, ParcelDeserializer())
            builder.create()
        }
    }
}